File: defmi/ledger.py
"""A ledger whose balances are commitments, not numbers.

The point of hiding a request and then settling it in the clear is nothing, so
the settlement layer has to hold balances it cannot read. Pedersen commitments
give exactly the structure needed:

    transfer of a from A to B      B_A' = B_A / C_a ,  B_B' = B_B * C_a
    conservation                   the product of all balances is unchanged, by
                                   construction, with no proof required
    solvency                       proved per transfer: the amount is not
                                   negative and the payer's new balance is not
                                   negative

Accounts are opaque handles. The ledger never learns which legal entity or which
asset a handle stands for; that meaning is established elsewhere and attested by
the computing quorum. What the ledger enforces on its own is arithmetic:
nothing is created, nothing goes negative, and nothing settles twice.
"""

from __future__ import annotations

import hashlib
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey, Ed25519PublicKey,
)

import hashlib
from dataclasses import dataclass, field
from typing import Any, Mapping

from zk.commit import Pedersen, RangeProof, prove_bounded, verify_bounded
from zk.groups import Group

LEDGER_DOMAIN = b"QOMM:DEFMI:LEDGER:v1"

# Balances live in a declared range. The width has to be fixed in advance
# because a range proof is only meaningful against a published bound.
MAX_BALANCE = (1 << 40) - 1


def _tagged_context(group, context: bytes, tag) -> bytes:
    """Bind the tag into the transcript.

    The prover picks the tag, so the generator the range proofs are stated
    against is prover-chosen. No attack follows --- the verification equations
    already pin it --- but a challenge that does not cover a value the prover
    controls is the kind of gap worth closing while it is one line.
    """
    if tag is None:
        return context
    return context + b":tag:" + group.encode(tag.point)


@dataclass(frozen=True)
class TransferSecrets:
    """The three blindings a transfer produces, which are not interchangeable.

    ``amount_blinding`` opens the amount commitment against the tag the leg was
    proved under; ``payee_delta`` is what the receiving account adds to its own
    blinding, which differs by gamma*amount once a tag is in play; and
    ``remainder_blinding`` is the payer's balance from here on. Returning them
    as one number is how a tagged leg silently stops verifying.
    """

    amount_blinding: int
    payee_delta: int
    remainder_blinding: int


@dataclass(frozen=True)
class TransferProof:
    """What a payer must show to move value it will not reveal."""

    amount_commitment: Any
    amount_range: RangeProof
    remainder_commitment: Any
    remainder_range: RangeProof
    tag: Any = None                      # the blinded asset tag, when there is one


@dataclass(frozen=True)
class LedgerEntry:
    handle: bytes
    balance: Any            # Pedersen commitment
    sequence: int


class InsufficientProof(ValueError):
    pass


class ConfidentialLedger:
    """Opaque handles to committed balances, with an auditable conservation check."""

    def __init__(self, group: Group, key: Pedersen | None = None,
                 max_balance: int = MAX_BALANCE,
                 issuer: Ed25519PublicKey | None = None):
        self.group = group
        self.key = key or Pedersen(group, b"qomm:defmi:v1")
        self.max_balance = max_balance
        # Who is allowed to create balance. A ledger without one takes any
        # opening commitment as issued, which makes conservation a statement
        # about what was admitted rather than about what exists --- see
        # `open_account`.
        self.issuer = issuer
        self._accounts: dict[bytes, LedgerEntry] = {}
        self._minted = group.identity()      # running total of everything issued
        self._issued: set[bytes] = set()     # authorisations already used

    # --- issuance -------------------------------------------------------
    def issuance_body(self, handle: bytes, balance_commitment, nonce: bytes) -> bytes:
        """What an issuer signs to allow one opening balance to exist."""
        return hashlib.sha256(
            b"QOMM:DEFMI:ISSUE:v1"
            + len(handle).to_bytes(4, "big") + handle
            + self.group.encode(balance_commitment)
            + len(nonce).to_bytes(4, "big") + nonce).digest()

    def open_account(self, handle: bytes, balance_commitment,
                     authorisation: bytes | None = None,
                     nonce: bytes = b"") -> LedgerEntry:
        """Admit a balance, and say who was allowed to create it.

        This used to take any commitment and fold it straight into `minted`, so
        the conservation check said "nothing has been created or destroyed since
        the ledger accepted these" and not "only an issuer created anything".
        Anyone who could call this could mint. A ledger constructed with an
        issuer now requires that issuer's signature over the handle, the
        commitment and a nonce it has not seen before; one without an issuer
        still accepts anything and says so, because the test fixtures and the
        conservation measurements do not need an issuance model and should not
        pretend to have one.
        """
        if handle in self._accounts:
            raise ValueError("handle already open")
        if self.issuer is not None:
            if authorisation is None:
                raise InsufficientProof(
                    "this ledger has an issuer, so an opening balance needs its "
                    "authorisation")
            body = self.issuance_body(handle, balance_commitment, nonce)
            if body in self._issued:
                raise InsufficientProof("that issuance authorisation was already used")
            try:
                self.issuer.verify(authorisation, body)
            except InvalidSignature as bad:
                raise InsufficientProof(
                    "the opening balance is not signed by the issuer") from bad
            self._issued.add(body)
        entry = LedgerEntry(handle, balance_commitment, 0)
        self._accounts[handle] = entry
        self._minted = self.group.mul(self._minted, balance_commitment)
        return entry

    def balance(self, handle: bytes):
        return self._accounts[handle].balance

    def handles(self) -> tuple[bytes, ...]:
        return tuple(sorted(self._accounts))

    # --- the invariant the ledger can check without reading anything ----
    def total(self):
        total = self.group.identity()
        for entry in self._accounts.values():
            total = self.group.mul(total, entry.balance)
        return total

    def conserved(self) -> bool:
        """No value was created or destroyed since the accounts were opened."""
        return self.group.encode(self.total()) == self.group.encode(self._minted)

    # --- transfers ------------------------------------------------------
    def build_transfer(self, payer_balance: int, payer_blinding: int, amount: int,
                       context: bytes, tag: "BlindedTag | None" = None,
                       gamma: int = 0, amount_bounded: bool = False
                       ) -> tuple[TransferProof, TransferSecrets]:
        """Run by the payer, who is the only party that knows its own balance.

        With an asset tag the proofs are stated against H = A h^gamma instead of
        the base generator. The balance itself stays committed under the bare A,
        so the remainder proof is what ties the disguise to the real asset: no
        other tag lets the payer open balance / amount as a value in range.

        ``amount_bounded`` drops the range proof on the amount. That proof is not
        redundant in general --- without it a payer can send a negative amount,
        which drains the payee --- but it *is* redundant when the amount is tied
        to an instruction that already proved it in range, which is every
        instruction-driven settlement. Half the range proofs in a delivery
        versus payment are this one, so the caller has to say so explicitly
        rather than have it inferred.
        """
        key = self.key if tag is None else tag.key_for(self.key)
        context = _tagged_context(self.group, context, tag)
        if amount < 0:
            raise InsufficientProof("a transfer amount cannot be negative")
        if payer_balance - amount < 0:
            raise InsufficientProof(
                f"balance {payer_balance} cannot cover {amount}")
        amount_blinding = key.random_blinding()
        if amount_bounded:
            amount_commitment = key.commit(amount, amount_blinding)
            amount_range = None
        else:
            amount_commitment, amount_range, _ = prove_bounded(
                key, amount, amount_blinding, 0, self.max_balance, context + b":amt")
        remainder = payer_balance - amount
        # H^(v-q) h^t has to land on balance / amount, which costs gamma*v
        remainder_blinding = (payer_blinding - amount_blinding
                              - gamma * payer_balance) % self.group.order
        remainder_commitment, remainder_range, _ = prove_bounded(
            key, remainder, remainder_blinding, 0, self.max_balance, context + b":rem")
        proof = TransferProof(amount_commitment, amount_range,
                              remainder_commitment, remainder_range, tag)
        return proof, TransferSecrets(
            amount_blinding=amount_blinding,
            payee_delta=(gamma * amount + amount_blinding) % self.group.order,
            remainder_blinding=remainder_blinding)

    def check_transfer(self, payer: bytes, proof: TransferProof, context: bytes,
                       amount_bounded: bool = False) -> tuple[bool, str]:
        """Run by the ledger, which reads neither the balance nor the amount."""
        key = self.key if proof.tag is None else proof.tag.key_for(self.key)
        group = self.group
        context = _tagged_context(group, context, proof.tag)
        if payer not in self._accounts:
            raise KeyError("unknown payer handle")
        if amount_bounded:
            if proof.amount_range is not None:
                return False, "an externally bounded amount carries a stale range proof"
        elif proof.amount_range is None:
            return False, "the amount carries no range proof"
        elif not verify_bounded(key, proof.amount_commitment, proof.amount_range,
                                0, self.max_balance, context + b":amt"):
            return False, "amount not shown to be within the ledger range"
        if not verify_bounded(key, proof.remainder_commitment, proof.remainder_range,
                              0, self.max_balance, context + b":rem"):
            return False, "payer would be left with a negative balance"
        # the remainder must be exactly balance - amount, which the homomorphism
        # settles without anyone opening either value
        expected = group.mul(self._accounts[payer].balance,
                             group.neg(proof.amount_commitment))
        if group.encode(expected) != group.encode(proof.remainder_commitment):
            return False, "remainder does not equal balance minus amount"
        return True, "ok"

    def apply_transfer(self, payer: bytes, payee: bytes, proof: TransferProof) -> None:
        """Only called once both legs of a settlement have been checked."""
        group = self.group
        payer_entry = self._accounts[payer]
        payee_entry = self._accounts[payee]
        self._accounts[payer] = LedgerEntry(
            payer, proof.remainder_commitment, payer_entry.sequence + 1)
        self._accounts[payee] = LedgerEntry(
            payee, group.mul(payee_entry.balance, proof.amount_commitment),
            payee_entry.sequence + 1)

    def snapshot(self) -> bytes:
        """A digest of every handle and balance, for receipts and audit."""
        digest = hashlib.sha256(LEDGER_DOMAIN)
        for handle in sorted(self._accounts):
            entry = self._accounts[handle]
            digest.update(handle)
            digest.update(self.group.encode(entry.balance))
            digest.update(entry.sequence.to_bytes(8, "big"))
        return digest.digest()

File: rust/qomm-defmi/src/ledger.rs
//! A ledger whose balances are commitments.
//!
//! What it can check without opening anything: value is neither created nor
//! destroyed, because the product of all balances equals the product at
//! issuance; and no balance goes negative, by a range proof on the difference.
//! What it cannot check, and does not try to, is whether the price was fair.
//! That meaning was established by the computing quorum and is carried by the
//! signature on the instruction.

use bulletproofs::{BulletproofGens, RangeProof};
use curve25519_dalek::ristretto::{CompressedRistretto, RistrettoPoint};
use curve25519_dalek::scalar::Scalar;
use curve25519_dalek::traits::Identity;
use merlin::Transcript;
use qomm_zk::pedersen::Pedersen;
use rand_core::{CryptoRng, RngCore};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;

use crate::assets::BlindedTag;
use qomm_zk::adaptor::{self, Signature};

pub struct Transfer {
    pub amount_commitment: RistrettoPoint,
    /// Absent when the amount is pinned by an instruction that already proved
    /// it in range. Half the range proofs in a settlement are this one, so the
    /// ledger requires the caller to say so rather than inferring it.
    pub amount_range: Option<(RangeProof, CompressedRistretto)>,
    pub remainder_commitment: RistrettoPoint,
    pub remainder_range: RangeProof,
    pub tag: Option<BlindedTag>,
}

pub struct TransferSecrets {
    /// Opens the amount commitment against the tag the leg was proved under.
    pub amount_blinding: Scalar,
    /// What the receiving account adds to its own blinding; differs by
    /// gamma*amount once a tag is in play. Returning one number for both is how
    /// a tagged leg silently stops verifying.
    pub payee_delta: Scalar,
    pub remainder_blinding: Scalar,
}

/// A transfer that has been checked and whose funds have left the payer, but
/// which has not reached the payee yet.
///
/// This is what makes a settlement across two ledgers possible at all. Neither
/// ledger can see the other, so neither can decide alone whether the pair went
/// through; what each *can* do is take the money out of reach and wait. The
/// amount stays on the ledger's books the whole time --- it is in neither
/// account, so conservation has to count it explicitly, which is the one place
/// this could silently create or destroy value.
pub struct Pending {
    pub payer: Vec<u8>,
    pub payee: Vec<u8>,
    pub amount_commitment: RistrettoPoint,
    /// Past this, the payer may take it back. A ledger does not read a clock;
    /// the caller supplies the time it is judging against, so the same code
    /// runs against a block height, a slot number or a wall clock.
    pub deadline: u64,
}

pub struct Ledger {
    pub key: Pedersen,
    pub bits: usize,
    bp_gens: BulletproofGens,
    accounts: BTreeMap<Vec<u8>, RistrettoPoint>,
    pending: BTreeMap<Vec<u8>, Pending>,
    minted: RistrettoPoint,
}

impl Ledger {
    pub fn new(key: Pedersen, bits: usize) -> Self {
        Ledger {
            key, bits,
            bp_gens: BulletproofGens::new(bits, 1),
            accounts: BTreeMap::new(),
            pending: BTreeMap::new(),
            minted: RistrettoPoint::identity(),
        }
    }

    pub fn open(&mut self, handle: &[u8], balance: RistrettoPoint) {
        assert!(!self.accounts.contains_key(handle), "handle already open");
        self.accounts.insert(handle.to_vec(), balance);
        self.minted += balance;
    }

    pub fn balance(&self, handle: &[u8]) -> Option<&RistrettoPoint> {
        self.accounts.get(handle)
    }

    pub fn handles(&self) -> Vec<Vec<u8>> { self.accounts.keys().cloned().collect() }

    /// No proof and no opening: the two products simply have to agree.
    ///
    /// Money held in escrow is in no account and is still on the books, so it
    /// is counted here. Leaving it out would make every prepared leg look like
    /// value destroyed and every commit like value created.
    pub fn conserved(&self) -> bool {
        let held: RistrettoPoint = self.pending.values()
            .map(|p| p.amount_commitment).sum();
        self.accounts.values().sum::<RistrettoPoint>() + held == self.minted
    }

    pub fn pending(&self, leg: &[u8]) -> Option<&Pending> { self.pending.get(leg) }

    pub fn pending_legs(&self) -> Vec<Vec<u8>> { self.pending.keys().cloned().collect() }

    /// Check a transfer and take the amount out of the payer's reach, without
    /// giving it to the payee.
    ///
    /// `leg` names the escrow. It is the caller's identifier and it is what the
    /// signature that later releases the money is made over, so it has to be
    /// unique on this ledger --- a reused name is a second claim on the first
    /// escrow, which is refused here rather than left to the caller.
    pub fn prepare_transfer(
        &mut self, leg: &[u8], payer: &[u8], payee: &[u8], transfer: &Transfer,
        context: &[u8], amount_bounded: bool, deadline: u64,
    ) -> Result<(), &'static str> {
        if self.pending.contains_key(leg) {
            return Err("that leg is already prepared");
        }
        if !self.accounts.contains_key(payee) {
            return Err("unknown payee handle");
        }
        self.check_transfer(payer, transfer, context, amount_bounded)?;
        self.accounts.insert(payer.to_vec(), transfer.remainder_commitment);
        self.pending.insert(leg.to_vec(), Pending {
            payer: payer.to_vec(),
            payee: payee.to_vec(),
            amount_commitment: transfer.amount_commitment,
            deadline,
        });
        Ok(())
    }

    /// Release an escrow to its payee.
    ///
    /// The ledger checks an ordinary signature over the leg's name under the
    /// payer's key. It knows nothing about adaptors or about the other ledger;
    /// what makes the pair atomic happens outside, and this stays a signature
    /// check so that it can.
    pub fn commit_pending(
        &mut self, leg: &[u8], payer_key: &RistrettoPoint,
        release: &Signature, now: u64,
    ) -> Result<(), &'static str> {
        let held = self.pending.get(leg).ok_or("no such prepared leg")?;
        if now > held.deadline {
            return Err("the escrow has expired");
        }
        if !adaptor::verify(payer_key, leg, release) {
            return Err("the release is not signed for this leg");
        }
        let held = self.pending.remove(leg).expect("just looked it up");
        let credited = self.accounts[&held.payee] + held.amount_commitment;
        self.accounts.insert(held.payee, credited);
        Ok(())
    }

    /// Give an expired escrow back to its payer.
    ///
    /// No signature: the payer already owned this money and the deadline is the
    /// whole authority. Requiring one would strand funds whenever the payer
    /// lost a key, which is the failure this branch exists to prevent.
    pub fn unwind_pending(&mut self, leg: &[u8], now: u64) -> Result<(), &'static str> {
        let held = self.pending.get(leg).ok_or("no such prepared leg")?;
        if now <= held.deadline {
            return Err("the escrow has not expired");
        }
        let held = self.pending.remove(leg).expect("just looked it up");
        let restored = self.accounts[&held.payer] + held.amount_commitment;
        self.accounts.insert(held.payer, restored);
        Ok(())
    }

    fn gens(&self, tag: Option<&BlindedTag>) -> bulletproofs::PedersenGens {
        match tag {
            Some(t) => t.gens(&self.key),
            None => bulletproofs::PedersenGens { B: self.key.g, B_blinding: self.key.h },
        }
    }

    fn context(&self, context: &[u8], tag: Option<&BlindedTag>) -> Vec<u8> {
        // The tag is prover-chosen, so it belongs in the transcript. The
        // verification equations already pin it; a challenge that does not
        // cover a value the prover controls is a gap worth closing anyway.
        let mut out = context.to_vec();
        if let Some(t) = tag {
            out.extend_from_slice(b":tag:");
            out.extend_from_slice(t.point.compress().as_bytes());
        }
        out
    }

    /// Run by the payer, the only party that knows its own balance.
    #[allow(clippy::too_many_arguments)]
    pub fn build_transfer<R: RngCore + CryptoRng>(
        &self, payer_balance: u64, payer_blinding: &Scalar, amount: u64,
        context: &[u8], tag: Option<&BlindedTag>, gamma: &Scalar,
        amount_bounded: bool, rng: &mut R,
    ) -> Result<(Transfer, TransferSecrets), &'static str> {
        if payer_balance < amount {
            return Err("balance cannot cover the amount");
        }
        let gens = self.gens(tag);
        let ctx = self.context(context, tag);
        let amount_blinding = Scalar::random(rng);

        let amount_range = if amount_bounded { None } else {
            let mut t = Transcript::new(b"qomm:defmi:amt");
            t.append_message(b"ctx", &ctx);
            let (proof, commitments) = RangeProof::prove_multiple(
                &self.bp_gens, &gens, &mut t, &[amount], &[amount_blinding], self.bits)
                .map_err(|_| "amount range proof failed")?;
            Some((proof, commitments[0]))
        };
        let amount_commitment = gens.commit(Scalar::from(amount), amount_blinding);

        // H^(v-q) h^t has to land on balance / amount, which costs gamma*v
        let remainder = payer_balance - amount;
        let remainder_blinding =
            payer_blinding - amount_blinding - gamma * Scalar::from(payer_balance);
        let mut t = Transcript::new(b"qomm:defmi:rem");
        t.append_message(b"ctx", &ctx);
        let (remainder_range, remainder_commitments) = RangeProof::prove_multiple(
            &self.bp_gens, &gens, &mut t, &[remainder], &[remainder_blinding], self.bits)
            .map_err(|_| "remainder range proof failed")?;

        Ok((
            Transfer {
                amount_commitment,
                amount_range,
                remainder_commitment: remainder_commitments[0].decompress()
                    .ok_or("bad remainder commitment")?,
                remainder_range,
                tag: tag.cloned(),
            },
            TransferSecrets {
                amount_blinding,
                payee_delta: gamma * Scalar::from(amount) + amount_blinding,
                remainder_blinding,
            },
        ))
    }

    /// Run by the ledger, which reads neither the balance nor the amount.
    pub fn check_transfer(
        &self, payer: &[u8], transfer: &Transfer, context: &[u8], amount_bounded: bool,
    ) -> Result<(), &'static str> {
        let balance = *self.accounts.get(payer).ok_or("unknown payer handle")?;
        let gens = self.gens(transfer.tag.as_ref());
        let ctx = self.context(context, transfer.tag.as_ref());

        match (&transfer.amount_range, amount_bounded) {
            (Some(_), true) => return Err("an externally bounded amount carries a stale range proof"),
            (None, false) => return Err("the amount carries no range proof"),
            (Some((proof, commitment)), false) => {
                if *commitment != transfer.amount_commitment.compress() {
                    return Err("the range proof is about a different amount");
                }
                let mut t = Transcript::new(b"qomm:defmi:amt");
                t.append_message(b"ctx", &ctx);
                proof.verify_multiple(&self.bp_gens, &gens, &mut t,
                                      std::slice::from_ref(commitment), self.bits)
                    .map_err(|_| "amount not shown to be within the ledger range")?;
            }
            (None, true) => {}
        }

        let mut t = Transcript::new(b"qomm:defmi:rem");
        t.append_message(b"ctx", &ctx);
        transfer.remainder_range
            .verify_multiple(&self.bp_gens, &gens, &mut t,
                             &[transfer.remainder_commitment.compress()], self.bits)
            .map_err(|_| "payer would be left with a negative balance")?;

        // the remainder must be exactly balance - amount, which the
        // homomorphism settles without anyone opening either value
        if balance - transfer.amount_commitment != transfer.remainder_commitment {
            return Err("remainder does not equal balance minus amount");
        }
        Ok(())
    }

    /// Only called once every leg of a settlement has been checked.
    pub fn apply_transfer(&mut self, payer: &[u8], payee: &[u8], transfer: &Transfer) {
        self.accounts.insert(payer.to_vec(), transfer.remainder_commitment);
        let credited = self.accounts[payee] + transfer.amount_commitment;
        self.accounts.insert(payee.to_vec(), credited);
    }

    pub fn snapshot(&self) -> [u8; 32] {
        let mut hasher = Sha256::new();
        hasher.update(b"QOMM:DEFMI:LEDGER:v1");
        for (handle, balance) in &self.accounts {
            hasher.update((handle.len() as u32).to_be_bytes());
            hasher.update(handle);
            hasher.update(balance.compress().as_bytes());
        }
        hasher.finalize().into()
    }
}
